### PR TITLE
feat: UIElements Welcome Window

### PR DESCRIPTION
**Pull request by uweenukr**
UIElements version of #508

### CODE REVIEW

File: Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+                //return nothing if path wasn't found
+                return "";
+            }
+        }
+
+        //get the start up key
+        private static string firstStartUpKey = string.Empty;
+
+        //get the icon texture
+        private static Texture2D MirrorIcon = null;
```
Comment (paulpach) [inline Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs:103]:
```suggestion
        private static Texture2D MirrorIcon;
```

File: Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        //window size of the welcome screen
+        private static Vector2 windowSize = new Vector2(500, 600);
+
+        //returns the path to the Mirror folder (ex. Assets/Mirror)
+        private static string MirrorPath
+        {
+            get
+            {
+                //get an array of results based on the search
+                string[] results = AssetDatabase.FindAssets("", new[] { "Assets" });
```
Comment (paulpach) [inline Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs:82]: I don't think this works with UPM.

A better way is to make this property not static,  call `MonoScript.FromScriptableObject(this)` to get the script asset and then call `AssetDatabase.GetAssetPath()` to get the path to this script.

File: Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+            var uxml = AssetDatabase.LoadAssetAtPath<VisualTreeAsset>(MirrorPath + "/Editor/WelcomeWindow/WelcomeWindow.uxml");
+            var uss = AssetDatabase.LoadAssetAtPath<StyleSheet>(MirrorPath + "/Editor/WelcomeWindow/WelcomeWindow.uss");
+
+            //Load the descriptions
+
+            uxml.CloneTree(root);
+            root.styleSheets.Add(uss);
+
+            //set the icon image
+            var icon = root.Q<Image>("Icon");
```
Comment (paulpach) [inline Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs:195]: Can't this be done in the xml instead?

File: Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+            //get the header
+            var header = root.Q<Label>("Header");
+            //get the description
+            var description = root.Q<Label>("Description");
+            //get the redirect button
+            var redirectButton = root.Q<Button>("Redirect");
+
+            //init the content in the right column
+            header.text = GetPageData(EPageDataType.header).ToString();
```
Comment (paulpach) [inline Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs:210]: Can't this be done in the xml?

File: Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.uxml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+	<!-- Title text -->
+	<engine:Label name="Title" text="Welcome to MirrorNG!"/>
+
+	<!-- Display the page buttons and the descriptions -->
+	<engine:Box name ="ColumnContainer">
+
+		<!-- Display the page buttons -->
+		<engine:Box name="LeftColumnBox">
+			<engine:Button name="Button0"/>
```
Comment (paulpach) [inline Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.uxml:29]: Can this have a more meaningful name?

**Review by paulpach — CHANGES_REQUESTED** [review]
Move all the presentation stuff in the xml,  such as all the text and icon.

The code should just add the listeners to the buttons.

File: Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+        #region Urls
+
+        private const string welcomePageUrl = "https://mirrorng.github.io/MirrorNG/";
+        private const string quickStartUrl = "https://mirrorng.github.io/MirrorNG/Articles/Guides/CommunityGuides/MirrorQuickStartGuide/index.html";
+        private const string changelogUrl = "https://github.com/MirrorNG/MirrorNG/commits/master";
+        private const string bestPracticesUrl = "https://mirrorng.github.io/MirrorNG/Articles/Guides/BestPractices.html";
+        private const string faqUrl = "https://mirrorng.github.io/MirrorNG/Articles/Guides/FAQ.html";
+        private const string sponsorUrl = "";
+        private const string discordInviteUrl = "https://discord.gg/N9QVxbM";
```
Comment (paulpach) [inline Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs:69]:
```suggestion
```

Should be able to get rid of all of these and just add them in the XML file directly.

**Review by paulpach — CHANGES_REQUESTED** [review]

**Comment by paulpach** [comment]
you seem to have accidentally changed the permissions of some .meta files

**Comment by uweenukr** [comment]
opps :(

File: Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs (hunk 6 above)
Comment (erikastaroza) [reply Assets/Mirror/Editor/WelcomeWindow/WelcomeWindow.cs:69]: This won't work for the page switching.

**Comment by paulpach** [comment]
Screen Shot 2020-12-26 at 11 22 59 AM
White text looks unreadable

**Comment by paulpach** [comment]
Fixed theme support.  This is what it looks like now:
Screen Shot 2020-12-26 at 12 56 10 PM
Screen Shot 2020-12-26 at 12 56 36 PM

**Pull request merged**